### PR TITLE
Add PodMonitor resources to the Helm chart

### LINKED ISSUE #6596: Prometheus-Operator PodMonitor Support
### What problem are you trying to solve?

I have powerful Prometheus setup deployed with prometheus-operator, extended with Thanos.
I want to bring my own Prometheus without changing prometheus config. But do it in prometheus-operator way.

### How should the problem be solved?

The Prometheus-Operator provided [PodMonitor](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#podmonitor) resource.  It help prometheus to discover targets (pods) by LabelSelector. 
(Under the hood, it translate and append the monitor CRD into prometheus config. Then reload Prometheus config.)

Since linkerd will label the injected pods. A single PodMonitor will be enough for all sidecars to be discovered & scrape by Prometheus.

### Any alternatives you've considered?
If we don't want this built in at least added the tutorial to the doc.

### How would users interact with this feature?
Some command that generate a `PodMonitor` for user to install.

### PR DESCRIPTION
**Pull request by jaygridley**
Add PodMonitor resources to the Helm chart

With an external Prometheus setup installed using prometheus-operator the Prometheus instance scraping can be configured using Service/PodMonitor resources.

By adding PodMonitor resource into Linkerd Helm chart we can mimic the configuration of bundled Prometheus, see https://github.com/linkerd/linkerd2/blob/main/viz/charts/linkerd-viz/templates/prometheus.yaml#L47-L151, that comes with linkerd-viz extension. The PodMonitor resources are based on https://github.com/linkerd/website/issues/853#issuecomment-913234295 which are proven to be working. The only problem we face is that bundled Grafana charts will need to look at different jobs when querying metrics.

When enabled by `podMonitor.enabled` value in the Helm chart, PodMonitor for Linkerd resources should be installed alongside the Linkerd and Linkerd metrics should be present in the Prometheus.

Fixes #6596

Signed-off-by: Martin Odstrcilik <martin.odstrcilik@gmail.com>

### CODE REVIEW

**Review by SuitableProduce — COMMENTED** [review]
Looks good. Consider a merge?

**Comment by junnplus** [comment]
Is there any progress?

File: charts/linkerd-control-plane/templates/podmonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  podMetricsEndpoints:
+    - interval: {{ $podMonitor.scrapeInterval }}
+      scrapeTimeout: {{ $podMonitor.scrapeTimeout }}
+      relabelings:
+        - sourceLabels:
+            - __meta_kubernetes_pod_container_name
+            - __meta_kubernetes_pod_container_port_name
+            - __meta_kubernetes_pod_label_linkerd_io_control_plane_ns
+          action: keep
+          regex: ^linkerd-proxy;linkerd-admin;linkerd$
```
Comment (alpeb) [inline charts/linkerd-control-plane/templates/podmonitor.yaml:94]:
```suggestion
          regex: ^linkerd-proxy;linkerd-admin;{{.Release.Namespace}}$
```

**Review by alpeb — COMMENTED** [review]
Thanks for this @jaygridley and sorry for the late review!
This looks good to me, just needs to be brought up to date against `main`.

File: charts/linkerd-control-plane/templates/podmonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+{{ if and $podMonitor.enabled .Values.podMonitor.proxy.enabled -}}
+---
+###
+### Prometheus Operator PodMonitor Linkerd data-plane
+###
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: "linkerd-proxy"
+  {{ include "partials.namespace" . }}
```
Comment (alpeb) [inline charts/linkerd-control-plane/templates/podmonitor.yaml:75]: This still needs to be replaced with `namespace: {{ .Release.Namespace }}`

**Review by alpeb — COMMENTED** [review]
While testing this I realized support for CLI is still missing, e.g. to be able to do something like
```
bin/linkerd install --set podMonitor.enabled=true
```
For this, you need to add a path to `podmonitor.yaml` into `templatesControlPlane` in `cli/cmd/install.go`, and a new entry in `Values` at `pkg/charts/linkerd2/values.go`. That entry could just be
```
PodMonitor                   map[string]interface{} `json:"podMonitor"``
```

File: pkg/charts/linkerd2/values.go [hunk 3]
```diff
@@ -138,6 +139,32 @@ type (
...
 	}
 
+	// PodMonitor contains the fields to configure the Prometheus Operator `PodMonitor`
+	PodMonitor struct {
+		Enabled        bool                     `json:"enabled"`
+		ScrapeInterval string                   `json:"scrapeInterval"`
+		ScrapeTimeout  string                   `json:"scrapeTimeout"`
+		Controller     *PodMonitorController    `json:"controller"`
+		ServiceMirror  *PodMonitorServiceMirror `json:"serviceMirror"`
+		Proxy          *PodMonitorProxy         `json:"proxy"`
```
Comment (alpeb) [inline pkg/charts/linkerd2/values.go:149]: Given these are equal, I think you could point to the same struct.

File: pkg/charts/linkerd2/values.go [hunk 4]
```diff
@@ -138,6 +139,32 @@ type (
...
+	}
+
+	// PodMonitorController contains the fields to configure the Prometheus Operator `PodMonitor` for the control-plane
+	PodMonitorController struct {
+		Enabled           bool   `json:"enabled"`
+		NamespaceSelector string `json:"namespaceSelector"`
```
[the remark this answers is not in the document]
Comment (jaygridley) [reply pkg/charts/linkerd2/values.go:155]: `NamespaceSelector` should be a `string` as it is evaluated using Helm `tpl` func

[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/charts/linkerd2/values.go:155]: Ah I didn't notice the usage of `tpl`, that's cool :+1:

**Comment by alpeb** [comment]
As a last step, please run `go test ./... -update` to update the golden test files as per the `values_test.go` changes.

**Comment by alpeb** [comment]
Also, if you check the diff after running `go test ./... -update` you'll see `podmonitor.yaml` is introducing an empty line in the output. Can you please check that out? It's gotta be something related to the `{{-` `-}}` space-suppressing tags.

File: charts/linkerd-control-plane/templates/podmonitor.yaml [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+          regex: __meta_kubernetes_pod_label_(.+)
+          replacement: __tmp_pod_label_$1
+        - action: labelmap
+          regex: __tmp_pod_label_linkerd_io_(.+)
+          replacement: __tmp_pod_label_$1
+        - action: labeldrop
+          regex: __tmp_pod_label_linkerd_io_(.+)
+        - action: labelmap
+          regex: __tmp_pod_label_(.+)
+{{- end -}}
```
Comment (alpeb) [inline charts/linkerd-control-plane/templates/podmonitor.yaml:122]: You can't suppress spaces on both sides or else the following resource won't have a proper `---` separator, as you can see by testing through the CLI `bin/linkerd install --set podMonitor.enabled=true`
```suggestion
{{- end }}
```

**Review by alpeb — APPROVED** [review]
Thanks @jaygridley ! This worked great for me :+1: 

This is how I tested:
```bash
# install kube-prometheus
$ git clone git@github.com:prometheus-operator/kube-prometheus.git
$ cd kube-prometheus
$ git checkout v0.11.0
$ kubectl apply --server-side -f manifests/setup
$ kubectl wait \
	--for condition=Established \
	--all CustomResourceDefinition \
	--namespace=monitoring
$ kubectl apply -f manifests/

# setup RBAC
$ cat << EOF | kubectl apply -f -
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: prometheus
rules:
- apiGroups: [""]
  resources:
  - nodes
  - nodes/metrics
  - services
  - endpoints
  - pods
  verbs: ["get", "list", "watch"]
- apiGroups: [""]
  resources:
  - configmaps
  verbs: ["get"]
- apiGroups:
  - networking.k8s.io
  resources:
  - ingresses
  verbs: ["get", "list", "watch"]
- nonResourceURLs: ["/metrics"]
  verbs: ["get"]
EOF

$ cat << EOF | kubectl apply -f -
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: prometheus
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: prometheus
subjects:
- kind: ServiceAccount
  name: prometheus-k8s
  namespace: monitoring
EOF

# delete network policy to be able to access the prom API
$ kubectl -n monitoring delete networkpolicies.networking.k8s.io prometheus-k8s

# rollout prom pods
$ kubectl -n monitoring rollout restart statefulset prometheus-k8s

# install linkerd
$ bin/linkerd install --crds | k apply -f -
$ bin/linkerd install --set podMonitor.enabled=true | k apply -f -

# install linkerd viz pointing to the operator's prom
$ bin/linkerd viz install --set prometheus.enabled=false --set prometheusUrl=prometheus-k8s.monitoring:9090 | k apply -f -

# checkout all the nice stats
$ bin/linkerd viz stat -A deploy
```

**Review by adleong — APPROVED** [review]

**Pull request merged**